### ISSUE TITLE
Add tests for broken stories

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#3378]
## Feature Description

We use Storybook as the foundation of our visual regression tests but we also have many more stories than we have VRT scenarios. Currently VRT is the only test we have in place to prevent Storybook from being broken, but it only catches stories that have corresponding VRT scenarios which are almost entirely all legacy stories. It doesn't make sense to enable VRT for every story, but we should at least still be able to require that no stories are broken which is a problem that keeps coming up. 

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new `test:storybook` command should be created, which when run will check every story to see if it's broken or not
    * The command should list all broken stories by default although an option to fail on the first one found would be nice
    * If a broken story is found, the command should return with a non-zero exit code
    * Any new test scripts/config/files should go in a new `tests/storybook` directory
    * A broken story is simply one that results in a `sb-show-errordisplay` class on the `body` element in the preview iframe – this frame should always be loaded directly (like it is in VRT) rather than in the larger view we normally use
* The new command should be integrated into our current GitHub actions so that broken stories cause the Storybook workflow to fail 

## Implementation Brief
* Install the `@storybook/addon-storyshots-puppeteer` package as dev-dependency along with `puppeteer` as per the documentation. https://storybook.js.org/addons/@storybook/addon-storyshots-puppeteer
* Create `.storybook/Stories.test.js` which contains the code sample in the "Specifying the storybook URL" section of the documentation, using the  local static build of storybook found in `/dist` via `file://`.
* The `setupTimeout` and `testTimeout` values are `15s` by default and this can be reduced to `5s` as per the documentation.
* Update the `parameters` object in `.storybook/preview.js` to include the asynchronous `puppeteerTest` function (refer to documentation), where the function checks whether the `body` of the rendered story contains the `sb-show-errordisplay` class name. For e.g:
    ```js
    expect( await page.$eval( 'body', ( el ) => el.classList.contains( 'sb-show-errordisplay' ) ) ).toBe( false );
    ```
* Update `package.json` to include a new command, `test:storybook` which runs:
    ```
    npm run test:js -- .storybook/Stories.test.js
    ```
* Using `.github/workflows/storybook.yml`,
    * Add a new job `test-storybook` which uses the code from the existing `build-storybook` job with the following changes:
        * Instead of running `npm run build:storybook`, run the `npm run test:storybook` command.
        * Remove `Upload artifacts` section.

### Test Coverage
* No further tests to be added.

### Visual Regression Changes
* N/A

## QA Brief

This should maybe be a QA:Eng ticket as there are no user-facing changes

* Check that all checks are passing on GitHub

## Changelog entry

* N/A

**Comment by aaemnnosttv** [comment]
@felixarntz I think this one is probably overdue 😄  Anything to add/update before IB?
cc: @tofumatt @eugene-manuilov

**Comment by felixarntz** [comment]
👏 Let's do it, lgtm!

**Comment by eugene-manuilov** [comment]
> A broken story is simply one that results in a sb-show-errordisplay class on the body element in the preview iframe – this frame should always be loaded directly (like it is in VRT) rather than in the larger view we normally use

@aaemnnosttv sometimes widgets become broken in a different way. They don't throw an error, they display an incorrect state instead. Not sure how we can catch it without VRT :thinking:

**Comment by aaemnnosttv** [comment]
> @aaemnnosttv sometimes widgets become broken in a different way. They don't throw an error, they display an incorrect state instead. Not sure how we can catch it without VRT 🤔

@eugene-manuilov Sure – anything not covered by VRT will be possible to be incorrect which is also not good, but for the purposes of this issue, the intention is just to ensure we at least catch the glaringly obvious broken ones which have been missed many times before. We can start with this and see how we can best evolve the stability of our stories going forward.

**Comment by eugene-manuilov** [comment]
Ok, sounds good to me. :+1:

**Comment by tofumatt** [comment]
IB looks good, I'm excited for this! 👍🏻 

IB ✅

**Comment by aaemnnosttv** [comment]
@felixarntz it would be great if we could get this into the next sprint 🙂

**Comment by danielgent** [comment]
Running `"test:storybook": "npm run test:js -- .storybook/Stories.test.js",` doesn't run any tests

It just gives `Pattern: .storybook/Stories.test.js - 0 matches`


See https://github.com/google/site-kit-wp/compare/infrastructure/3378-Add-tests-for-broken-stories

**Comment by danielgent** [comment]
I've moved it to `assets/js/modules/Stories.test.js` for now, as this is picked up by the Jest config :+1:

**Comment by danielgent** [comment]
I just had to add `'<rootDir>/.storybook/?(*.)test.js',` under `testMatch` in `jest.config.js` to pick up `.storybook/Stories.test.js"`

**Comment by danielgent** [comment]
Amazing! this totally works

![works](https://user-images.githubusercontent.com/8496063/128011889-b4d4f2f5-7b4d-40c7-9fe9-96308926bc2e.png)


Now the problem is how to make a PR that will fail because we actually have broken stories @aaemnnosttv 

I can comment them out *or* we can fix them first? :smile:

**Comment by danielgent** [comment]
Example broken story https://google.github.io/site-kit-wp/storybook/develop/?path=/story/analytics-module-settings--no-tag-gtm-property-w-access

I'm sure I've created a ticket to fix these


I'll make a PR now anyway so we can see it failing

**Comment by danielgent** [comment]
When running on a Github workflow though the steps in the IB aren't sufficient as we haven't built storybook and put it into a local `dist` folder

*All* the tests fail with

`ERR! Error when connecting to file:///home/runner/work/site-kit-wp/site-kit-wp/dist/iframe.html?id=global--accordion, did you start or build the storybook first? A storybook instance should be running or a static version should be built when using puppeteer test feature.`

I'm going to try copying a few steps now
1. Downloading the built artifacts
2. Building storybook

**Comment by danielgent** [comment]
Downloading the artifacts didn't work (I get an error `'An Artifact with name "storybook-files" was not found.'`)

*But* building storybook first did work

![build first](https://user-images.githubusercontent.com/8496063/128015827-9cafae20-7222-40b9-90e8-9bf0fa502292.png)

Seems excessive to build twice though.  We could just combine the Build and Test storybook jobs @aaemnnosttv

**Comment by aaemnnosttv** [comment]
@danielgent it fails to download the artifact because the job needs to depend on the `build` job for that to be available. The `job.needs` is missing.

Regarding the broken stories, I mentioned on the others that it didn't make sense to fix them again and again until this issue was done so, it's somewhat out of scope but unless they're substantially too much effort let's just fix them as part of this PR.

**Comment by danielgent** [comment]
Sure thing Evan

So next I'll do 
1. Add `job.needs` so the storybook test step is dependent on the build step
2. Fix the broken stories if they aren't too difficult

Also I need to revisit my changes to `jest.config.js` as my changes to `testMatch` cause the `JS Test` step to fail :thinking:  (as now it's running `stories.test.js` )

**Comment by danielgent** [comment]
I'm not making any progress trying to fix these broken stories.  I'm finding the code very cryptic. Helper functions are passed to helper functions and the `registry` isn't making it all the way through

On this line https://github.com/google/site-kit-wp/blob/develop/stories/module-analytics-settings.stories.js#L48  
`Component( { callback, registry } ) {` - `registry` is null.  That's as far as I've got :joy: 

I don't see how `generateGTMAnalyticsPropertyStory` has ever passed in `registry` to the inner component

Honestly I would prefer to just get rid of this `generateGTMAnalyticsPropertyStory`. It's really hard to follow. This `generateGTMAnalyticsPropertyStory` just runs sets up the registry anyway, so I'll try extracting this and just running that code inline :sweat:

**Comment by aaemnnosttv** [comment]
@danielgent I agree some of the generator utilities around the old stories are a bit hard to follow.

>  I'll try extracting this and just running that code inline

Please don't do this just yet if it requires substantial changes. If needed we can always handle the broken parts in a separate issue. I'll take a look and see if I can figure out what's going on there.

**Comment by aaemnnosttv** [comment]
@danielgent I pushed a few commits which I think should fix things there. Let me know if you run into any other problems 👍

**Comment by danielgent** [comment]
All tests now passing :+1:   Moving to CR

**Comment by johnPhillips** [comment]
QA ✅
I can see that all tests are passing on Github.
